### PR TITLE
Add methods AddLinkAsync/RemoveLinkAsync/UpdateAsync/DeleteAsync to the Collection Fetcher Classes

### PR DESCRIPTION
**Pull request by Sridhar-MS**
- AddLinkAsync - Adds an entity reference to an entity collection (entity sets, collection-valued navigation properties)
- RemoveLinkAsync - Removes an entity reference from an entity collection (entity sets, collection-valued navigation properties)
- UpdateAsync - Update a single entity in an entity collection (entity sets, collection-valued navigation properties)
- DeleteAsync - Delete a single entity from an entity collection (entity sets, collection-valued navigation properties)

The data-classes from now will only have properties that hold an entity value. It does not send/receive any request/response to/from an OData service. Any operation requiring to talk to the OData service has to go through the Fetcher and Collection classes.

The data classes still explicitly implement the data-interface navigation properties. Ideally this is not required, just the explicit properties are sufficient. But due to a bug in 'Expand' scenario this is not yet removed in the generated proxies.

### CODE REVIEW

**Comment by Sridhar-MS** [comment]
The failing test has nothing to do with the changes in this PR. It seems to be a flaky test which needs to be fixed. I am working on it.

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+using Microsoft.Its.Recipes;
+using Microsoft.MockService;
+using Microsoft.MockService.Extensions.ODataV4;
+using Newtonsoft.Json.Linq;
+using Vipr.Core.CodeModel;
+using Xunit;
+
+namespace CSharpLiteWriterUnitTests
+{
+    public class Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method : EntityTestBase
```
Comment (piotrpMSFT) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs]: Why not derive from NavigationPropertyTestBase?

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    {
+        private MockService _mockedService;
+        private OdcmEntityClass _navPropertyClass;
+        private OdcmProperty _navProperty;
+        private System.Type _navPropertyFetcherType;
+        private System.Type _navPropertyCollectionFetcherType;
+        private System.Type _navPropertyConcreteType;
+        private EntityArtifacts _navPropertyEntity;
```
Comment (piotrpMSFT) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs]: NavPropertyTestBase will give you this EntityArtifacts.

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+         * 
+         * Spec - http://docs.oasis-open.org/odata/odata/v4.0/errata02/os/complete/part1-protocol/odata-v4.0-errata02-os-part1-protocol-complete.html#_Toc406398333
+         */
+
+        [Fact]
+        public void It_adds_a_link_to_a_collection_valued_navigation_property()
+        {
+            var entityKeyValues = Class.GetSampleKeyArguments().ToArray();
+            var propertyPath = Class.GetDefaultEntityPath(entityKeyValues) + "/" + _navProperty.Name;
+            var entity2KeyValues = _navPropertyClass.GetSampleKeyArguments().ToArray();
```
Comment (piotrpMSFT) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs]: relatedEntity?

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                    .SetupPostEntity(_navPropertyEntity, entity2KeyValues))
+            {
+
+                var context = _mockedService.GetDefaultContext(Model);
+                var collection = context.CreateCollection(_navPropertyCollectionFetcherType, _navPropertyConcreteType,
+                    propertyPath);
+                var sourceInstance = context.CreateConcrete(ConcreteType);
+                var targetInstance = context.CreateConcrete(_navPropertyConcreteType);
+                
+                collection.InvokeMethod<Task>("AddLinkAsync", new object[] { sourceInstance, targetInstance, true }).Wait();
```
Comment (piotrpMSFT) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs]: We should test both sides of the delay save. This should be fairly easy:
- add a config setting to MockService that makes it throw if it gets an unexpected request
- Once you invoke the delaySave version of AddLinkAsync, setup the MockService to expect the AddLink request
- execute the batch

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_DeleteAsync_Method.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            using (_mockedService = new MockService()
+                    .SetupPostEntity(TargetEntity, entityKeyValues))
+            {
+                var context = _mockedService.GetDefaultContext(Model);
+                var collection = context.CreateCollection(CollectionType, ConcreteType,
+                    Class.GetDefaultEntitySetPath());
+                var instance = context.CreateConcrete(ConcreteType);
+
+                //delay save when deleting
+                collection.InvokeMethod<Task>("DeleteAsync", new object[] { instance, true }).Wait();
```
Comment (piotrpMSFT) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_DeleteAsync_Method.cs]: Should test the batch operation

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_RemoveLinkAsync_Method.cs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+using Microsoft.Its.Recipes;
+using Microsoft.MockService;
+using Microsoft.MockService.Extensions.ODataV4;
+using Newtonsoft.Json.Linq;
+using Vipr.Core.CodeModel;
+using Xunit;
+
+namespace CSharpLiteWriterUnitTests
+{
+    public class Given_an_OdcmClass_Entity_Collection_RemoveLinkAsync_Method : EntityTestBase
```
Comment (piotrpMSFT) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_RemoveLinkAsync_Method.cs]: NavigationPropertyTestBase?

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_RemoveLinkAsync_Method.cs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+                    .SetupPostEntity(_navPropertyEntity, entity2KeyValues))
+            {
+
+                var context = _mockedService.GetDefaultContext(Model);
+                var collection = context.CreateCollection(_navPropertyCollectionFetcherType, _navPropertyConcreteType,
+                    propertyPath);
+                var sourceInstance = context.CreateConcrete(ConcreteType);
+                var targetInstance = context.CreateConcrete(_navPropertyConcreteType);
+                
+                collection.InvokeMethod<Task>("RemoveLinkAsync", new object[] { sourceInstance, targetInstance, true }).Wait();
```
Comment (piotrpMSFT) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_RemoveLinkAsync_Method.cs]: Should test the batch operation

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_UpdateAsync_Method.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+                    .SetupPostEntity(TargetEntity, entityKeyValues))
+            {
+                var context = _mockedService.GetDefaultContext(Model);
+                var collection = context.CreateCollection(CollectionType, ConcreteType, Class.GetDefaultEntitySetPath());
+                var instance = context.CreateConcrete(ConcreteType);
+
+                instance.SetPropertyValue(_structuralInstanceProperty.Name, Any.Word());
+
+                //delay save when updating
+                collection.InvokeMethod<Task>("UpdateAsync", new object[] { instance, true }).Wait();
```
Comment (piotrpMSFT) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_UpdateAsync_Method.cs]: Should test the batch operation

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Navigation_Property_Collection.cs [hunk 9]
```diff
@@ -70,23 +72,23 @@ public void The_Fetcher_class_exposes_a_readonly_CollectionInterface_property()
 
 
         [Fact]
-        public void The_Concrete_interface_exposes_a_readonly_IPagedCollectionOfConcreteInterface_property()
+        public void The_Concrete_interface_exposes_a_readonly_IListOfConcreteInterface_property()
```
Comment (piotrpMSFT) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Navigation_Property_Collection.cs]: Can you add a page to our wiki describing the breaking changes we're making?

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs (hunk 1 above)
Comment (Sridhar-MS) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs]: Fixed the test to derive from 'NavigationPropertyTestBase'. It was not easily discoverable before.

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs (hunk 2 above)
Comment (Sridhar-MS) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs]: Fixed the test to derive from 'NavigationPropertyTestBase'.

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs (hunk 3 above)
Comment (Sridhar-MS) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs]: Fixed.

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_RemoveLinkAsync_Method.cs (hunk 6 above)
Comment (Sridhar-MS) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_RemoveLinkAsync_Method.cs]: Fixed the test to derive from 'NavigationPropertyTestBase'.

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs (hunk 4 above)
Comment (Sridhar-MS) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_AddLinkAsync_Method.cs]: Add the batch test.

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_DeleteAsync_Method.cs (hunk 5 above)
Comment (Sridhar-MS) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_DeleteAsync_Method.cs]: Add the batch test.

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_RemoveLinkAsync_Method.cs (hunk 7 above)
Comment (Sridhar-MS) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_RemoveLinkAsync_Method.cs]: Add the batch test.

File: test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_UpdateAsync_Method.cs (hunk 8 above)
Comment (Sridhar-MS) [inline test/CSharpLiteWriterUnitTests/Given_an_OdcmClass_Entity_Collection_UpdateAsync_Method.cs]: Add the batch test.

**Pull request merged**